### PR TITLE
Backmerge block renaming fixes/refactors from 6.4 branch into Gutenberg trunk

### LINKED ISSUE #55338: Convert Rename Menu Item to standard menu item not Plugin
**Update**: when implementing this we can [draw on a bug fix](https://github.com/WordPress/gutenberg/pull/55604) which went directly into the `wp/6.4` branch.

As suggested by @jsnajdr let's make the "Rename" menu item into a default item within the menu as per `BlockLockMenuItem`:

https://github.com/WordPress/gutenberg/blob/ef21f207827e1ac9e23c307ca2a7efaf4366b6c6/packages/block-editor/src/components/block-settings-menu-controls/index.js#L83

We can then remove the filter thus avoiding any perf concerns with the Plugin-based approach.

> Alternatively, the rename UI doesn't need to be a plugin. It could be hardcoded directly into the Block Settings Menu, like, for example, `BlockLockMenuItem` is.

I think what you suggest above is the most pragmatic option for now. This could be actioned in tandem with the follow up in https://github.com/WordPress/gutenberg/issues/55194 to make metadata a default global attribute.

_Originally posted by @getdave in https://github.com/WordPress/gutenberg/issues/55250#issuecomment-1761099347_

### LINKED ISSUE #55643: Block Renaming Followups for 6.5
The following needs to be undertaken in the WP 6.**5** cycle.

- [x] - Backport [the refactor (bugfix) to the feature undertaken in the `wp/6.4` branch](https://github.com/WordPress/gutenberg/pull/55604#issuecomment-1779632656)
- [x] - [Make "metadata" a global block attribute](https://github.com/WordPress/gutenberg/issues/55194)
- [x] - https://github.com/WordPress/gutenberg/issues/55338
- [x] https://github.com/WordPress/gutenberg/issues/56155

_Originally posted by @getdave in https://github.com/WordPress/gutenberg/issues/54426#issuecomment-1780764881_

### PR DESCRIPTION
**Pull request by getdave**
## What?

Back merges the refactors that were undertaken directly in the `wp/6.4` branch to improve Block Renaming support.

Namely:

- https://github.com/WordPress/gutenberg/pull/55604
- Closes https://github.com/WordPress/gutenberg/issues/55338

Also fixes [this Issue report](https://github.com/WordPress/gutenberg/issues/55643#issuecomment-1812739942).

Part of https://github.com/WordPress/gutenberg/issues/55643

## Why?

The WP 6.4 branch got ahead of this branch due to the nature of some bugfixes that had to land _directly_ in the 6.4 branch because Gutenerg trunk had already progressed during the 6.4 Beta/RC cycle.

So now we bring Gutenberg back into line with the improvements in 6.4 whilst also retaining the improvements already in `trunk`.

## How?

- Cherry pick changes from 6.4 branch.
- Fix merge conflicts and ensure functionality already in Gutenberg `trunk` is restored.
- Improve e2e tests - I have tried to strike a balance between "coverage" and avoiding creating lots of e2e tests (which are inherently "expensive"). If there are any scenarios you think I've missed please say.

## Testing Instructions

- New Post
- Add blocks:
    - Group
    - Paragraph
    - Heading
    - Quote
- Toggle list view
- Click blocks options menu for each block in list view and check that Rename works as expected (can rename, remove the custom name...etc).
- Toggle Inspector Controls for each blocks and check that the rename control udner `Advanced` panel also behaves as expected and changes are synchronised to the List View.
- Now check the Rename option appears even when the block is not selected:
    - select Paragraph
    - click the block options menu for the `Heading` block (**note:** the block that is _not_ selected).
    - verify Rename menu item is shown and works as expected.
- Now verify Query Block behaviour
    - go to Post Template inside Query block add wrap contents in a Group block
    - click the block options menu for that Group and check you only see the `Rename` option _once_ in the menu.
- Now check the following blocks do _not_ show the rename control on either List View menu _or_ in Inspector Controls:
    - core/block
    - core/template-part
    - core/pattern
    - core/navigation

### CODE REVIEW

**Comment by getdave** [comment]
## Next Steps

- Rewrite tests to assert control is present on blocks where support is disabled.
- Add tests to assert control is not present on blocks where support is disabled.

**Note**: it might be preferable to simply register two dummy blocks with/without support and assert on those rather than use Core blocks.

**Comment by getdave** [comment]
Urgh I found a bug...

File: packages/block-editor/src/components/block-rename/rename-control.js [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+		( select ) => {
+			const { getBlockAttributes } = select( blockEditorStore );
+
+			const _metadata = getBlockAttributes( clientId )?.metadata;
+			return {
+				metadata: _metadata,
+			};
+		},
+		[ clientId ]
+	);
```
Comment (scruffian) [inline packages/block-editor/src/components/block-rename/rename-control.js:30]: Is there a reason to return an object, or is it just to match what we do elsewhere and make it easier to add more properties in the future?

```suggestion
	const metadata = useSelect(
		( select ) => {
			const { getBlockAttributes } = select( blockEditorStore );

			return getBlockAttributes( clientId )?.metadata;
		},
		[ clientId ]
	);
```

**Review by scruffian — APPROVED** [review]
Code looks good and tests well

File: packages/block-editor/src/components/block-rename/rename-control.js (hunk 1 above)
Comment (getdave) [reply packages/block-editor/src/components/block-rename/rename-control.js:30]: See https://jardasn.blog/2021/01/22/some-best-practices-for-using-useselect-from-wordpress-data/comment-page-1/#:~:text=%7D-,Prefer,-returning%20objects%20with

File: packages/block-editor/src/components/block-rename/use-block-rename.js [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+ * WordPress dependencies
+ */
+import { getBlockSupport } from '@wordpress/blocks';
+
+export default function useBlockRename( name ) {
+	const metaDataSupport = getBlockSupport( name, 'renaming', true );
+
+	const supportsBlockNaming = !! (
+		true === metaDataSupport || metaDataSupport?.name
+	);
```
Comment (getdave) [inline packages/block-editor/src/components/block-rename/use-block-rename.js]: I think this is a mistake.

Comment (getdave) [reply packages/block-editor/src/components/block-rename/use-block-rename.js]: We only need to check for renaming support.

**Comment by getdave** [comment]
I believe the [Puppeteer tests have actually passed](https://github.com/WordPress/gutenberg/actions/runs/6956738787) despite them not reporting to Github.

**Pull request merged**